### ISSUE TITLE
Anti-squatting System for gno.land namespaces

### ISSUE THREAD

**Issue by moul** [gnolang/gno#2727]
We need a generic anti-squatting system for `r/sys/users` to set up an automatic auction system. This will prevent squatting and ensure fair access. Initially, `r/sys/users` will remain invite-only.

How to Contribute: Submit a plan in plain English or pseudo code outlining the design of the anti-squatting system. Focus on creating a generic, importable library or realm. We will review the design before moving to full implementation.

Related: #2189 #2827

---

Bounty Size: XXL; expected maximum reward: $16,000.

[Find out more on the bounty program.](https://gno.land/contribute) If you participate with intention of receiving the bounty, you must agree to the [Bounty Terms and Conditions.](https://docs.google.com/document/d/1aXrZ6japdAykB5FLmHCCeBZTo-2tbZQHSQi79ITaTK0/edit) 

[More bounties](https://github.com/gnolang/gno/labels/bounty) | [Contributing guidelines](https://github.com/gnolang/gno/blob/master/CONTRIBUTING.md)

**Comment by miguelito766** [comment]
Hi,

I'm not sure i got all the points but here's what i propose :

**Goals :**
- Prevent Squatting
- Ensure Fair access

**Almost simplified Plan :**
- The user needs to authentificate and sign a noop contract in order to watch the auction  #This way you got a timestamp
    -  The user got a token for his connection (AuthToken) that match with noop contract
    
- The system could match the given time (lets say 5min of connection) then disconnect with the user by AuthToken

- The user needs to connect back to the auction and generate another Noop/Token

- The auction site count how many times a wallet has connected and block access according to the fairness.

_---_

As a user, I want to watch an auction -> I need to sign to get an AuthToken
As the operator i define that the auction site allow X AuthToken for this Auction

My AuthToken could be like "Credits" -> 1 credit = X time of watching the product

**Concept :**

The concept of this is to implement two things :
- Log authentificate that can be useful in some cases
- A temporary token concept :
    - Matching a TXT/ID by the Log authentificate
    OR
    - Send back a token that burn himself after X time
        - if you don't have the token = NO ACCESS

**Generic part :**
- Create a no-op contract / Logs contract
- Create a package with the concept of AuthToken
    - ACCESS to something
    - TIME given by the token
    
    
**To go further more**
- Marketing / showcases an app
- Make token that are less volatile (2h of Time, splitted by login/out) by pointing login and logout to count
    - This way could be a SubscribeToken

**Comment by moul** [comment]
Thanks for your proposal. I think there's a bit of a misunderstanding about the scope of the anti-squatting system we're aiming for. The goal is to keep a potential permissionless system (like an auction system or similar) free from squatting. This involves:

- Preventing Multiple Registrations: Limiting people from registering multiple things (usernames, DAOs, etc) to park them.
- Safeguarding Precious Names: Protecting important names like "system", "admin", ...
- (Not sure) Managing Conflicts: Providing a way for famous brands to recover usernames related to their trademarks.

The solution will likely involve designing processes and having moderation DAOs. While advanced patterns like signing noop messages aren't necessary, a broader approach could include some proof of humanity mechanisms. However, the issue is open to any proposals. I'm more interested in reviewing various ideas than limiting the spectrum of possibilities.

Looking forward to more suggestions!

**Comment by thinhnx-var** [comment]
# anti-squatting system idea - specs
In this proposal, I am thinking about things that possible with our chain, some ideas are inspired by other chains.
Lets start discussion:
### Value Attribute:
A good namespace (ns) should have itself value, at least with individual user. Then every namespaces should come with a `price` attribute. Whenever someone want to register a new ns, an auction session is opened. If there is only 1 participant, then the default price is applied. For determining this price attribute, and prevent spamming coins to get all the ns, I prefer using Auction.
### Auction format:
I prefer the `sealed bidding auction` format:
- 2 phases of commit: in the 1st phase - `bidding phase`, commit your `encoded-price`, which required a `salt` variable. After a duration, lets say about 1 day / week, the session runs to 2nd phase - `reveal phase` which participants need to commit their `salt` value, in oder to reveal the price. The argorithm to encode/decode the price from `encoded-price` will be public. But the `salt` value is secret, and during the session, only user who submit the bid knows. This way we can mark the joining point of a participant.
- This 2 phases format help us keeping the privacy of `sealed bidding`, it is simple and transperent.
- I also propose to make an UI DApps - service which helping user a hand on creating the `encode-price` with their `salt`, showing the current status of bidding sessions.
- There should be a lock-down mechanism: locking amount of participant's coin till the end, and locking the transferable attribute of a ns after the session ends.
### More anti-squatting:
- In the future, when the ecosystem can support, I am thinking about the Decentralize ID (DID) / KYC methodology, which can help us in many fields, and this anti-squatting system is a good usecase.


### data-source about sensitive namespaces (DAO + realm):
- In this spec, I think we need a `ns-data-source` which contains the sensitive namespace, like `system`, `bitcoin`, `messi` ... public by a realms, data is pushed into that realm from govDAO voting, or `oracle` source. I am still working on `oracle` in our chain, and believe that this idea is possible and make sense.
- Whenever a registering for these namespaces is triggered, an additional action is required. The action may be `voting`, `restricting`, `allowing`... (can define later)

### trademarks and conflictions (DAO + realm):
- In my opinion, we need a realm + DAO that let user registering their trademark, or iconic namespace, with evidences supplied.
- Whenever a confliction comes from sides, a DAO voting is required, depend on trademark is registered or not, we can start a new voting session that decide the owner of namespace.

In the early stage, I think we can try to implement the auction for sealed bidding idea, which required auction package, and service / dapp to interact with bidding session.

Maybe there are something I miss-understood, and I am looking forward to hearing from you, about anything from this proposal.
Lets point out what make sense and what does not.

**Comment by moul** [comment]
The anti-squatting system should be implemented at the same time or before the auction system, not after. If a robust anti-squatting system is not in place, we should use an invite-only system instead of an auction.

The anti-squatting system, once developed, can open the door to the next need: an auction system or a similar permissionless system.

**Comment by miguelito766** [comment]
Hello,

My attention have been caught by this sentence :

> * (Not sure) Managing Conflicts: Providing a way for famous brands to recover usernames related to their trademarks.

----

In that situation, maybe we could link the NS registration with the public DNS registrar that will act like a source of truth.
The owner of the domain needs to add a TXT record to his DNS Zone in order to verify his domain name.

This might come with limitation (or solution ?) because the NS will be "famousbrand.com, famousbrand.net...." instead of just "Famousbrand".

This way you can't avoid scam domains like "gno.xyz, gno.net..." but at least you can certify that you are "gno.land".

----

> * Preventing Multiple Registrations: Limiting people from registering multiple things (usernames, DAOs, etc) to park them.

Resolved by the way above. If your reserved domain haven't been claimed then its remain open to claim.

> * Safeguarding Precious Names: Protecting important names like "system", "admin", ...

First you'll have a full name "gno.land" then you might reserve and attribute "shorter name" by proof-of-humanity system, moderation DAO etc...

Exemple : If i certified Famousbrand.com and i'm the actual owner i can ask a "shorter name" to get second check verification and get my final name "Famousbrand".

---

So the process could be :

1) Registration with a DNS name
2) Certify that you own the DNS name (First verify level) -> Permissionless
3) Ask for shorter name (Second verify level) -> Proof-of-humanity


Thanks for reading

**Comment by moul** [comment]
I like the idea of having a longer name easily, and then a shorter one (+ alias?) after more verification.

When working on this PR, I had this idea in mind: https://github.com/gnolang/gno/pull/2486

**Comment by AnhVAR** [comment]
I think we can mix the following ideas: 
There will be two types of namespaces. 
1. The first type is an unverified namespace, where users can register for free. 

This namespace must follow the format: <User namespace>-<RandomString>.gno. This ensures that any user can register their own namespace, and anyone looking at it can recognize it as unverified due to the RandomString at the end.

3. The second type is a verified namespace managed by the core team. 

After users register the first type mentioned above, if they have good activities, a strong community, regular users, and can complete KYC (Know Your Customer) or KYB (Know Your Business) processes, the core team will provide an alias short namespace by removing the RandomString for the registered and qualified namespace of the first type.

**Comment by thinhnx-var** [comment]
After reading your comments, I am now thinking about the idea that we have a `longer namespace` easily, and have a `verified shorter` namespace which will be managed by invite-only approach.
In summary, my idea is:
1. Having a type of longer namespace in a format like: `yourbrand` + `random-fixed-length` string. The random-fixed-length will be used as `Unique-ID` to specific the namespace. ( e.g.: `gno-123abc`, `gno-abc321`...)
2. Having a type of shorter namespace. Lets say `yourbrand`. This namespace is invited by admin - voted by DAO ... (humanity-proof). (e.g.: `gno`)

Users can start develope their contracts in `type-1` namespace. And when they have enough users, have marketplace..., they will want to own their `type-2` namespace. We will find the way to support porting from `type-2` to `type-1` namespace. This will helps them in save current status of their contracts, users, data...

In this way, we can help people know which namespace is trusted (type 2), and which namespace is in `untrusted` state ( type 1). And the squatting problem will also no longer exist, because the squatting does not make sense anymore.

What do you think @moul  @AnhVAR

**Comment by miguelito766** [comment]
> I like the idea of having a longer name easily, and then a shorter one (+ alias?) after more verification.
> 
> When working on this PR, I had this idea in mind: #2486

If you go on the domain verification solution, some questions will remain :

What happen when you loose your domain name ?
What happen if you delete the TXT verification ?

When you make a TXT verification, you want to check (every 24h for example) if the record is still alive, than delete the "certify/validation" if record is not here.

Possible solution : 
- Auto-burn or Transfer the NS in order to put it again "Open-to-claim" ? -> Then the namespace will take back the wallet address ( gno.land/{p,r}/g1xxxx/ )

-----

Regarding Alias, maybe you could link your users wallets to a wallet, at the link process you ask for your Alias ("Moul") to be linked gno.land/{p,r}/gno.land/moul


----

My final proposition for antisquatting :

Situation :
I am the owner of gno.land domain name, i want to bring my team on gno.

Permisionless Process :
- I create a wallet for my domain # g1xxx
- I register my domain on gnoland -> gno.land
- I receive a token wich i need to put on my domain records -> I create a new TXT record with that token on my dns registrar
- After an auto-check i claimed the NS to my wallet # Link g1xxx - gno.land
----
- My teammates create there wallet # g1mmm
- They send an approval to the wallet with alias name (that will be linked to NS) # Link g1xxx - g1mmm

Proof-of-humanity Process :
(- I've made the permissionless process) # ?
- I'm now able to ask for a shorter NS

Warning with this solution is gas used is nobody has NS # (gno.land/{p,r}/g1xxxx/g1mmm) that make longer path

**Comment by moul** [comment]
Opened an issue regarding on r/sys/users (#2827), so we can focus the discussion on this issue exclusively about an anti-squatting system, ideally pluggable as a `p/` and not limited to `r/sys/users`.

The anti-squatting system shouldn't depend on DNS verification.

**Comment by thehowl** [comment]
@moul I set an XXL bounty on this one; if you think it's insufficient, we can bump it to the next level.

**Comment by moul** [comment]
This is a must-have for r/sys/users/v2, after the launch, but not for the first launch. In the meantime, we can postpone it.

Must have issue: #2827

**Comment by moul** [comment]
Linked with #3020

**Comment by sudhanshu112233shukla** [comment]
review my changes @moul

**Comment by Kouteki** [comment]
Related to https://github.com/gnolang/gno/issues/4937

**Comment by repairman29** [comment]
🎯 **PING RECEIVED!**

**@repairman29** has claimed this bounty via [Echeo](https://echeo.io).

**Match Score:** 85%  
**Status:** Ready to start working!

[View conversation in Echeo](https://echeo.io/messages) to coordinate.

**Comment by repairman29** [comment]
Hi! I'm interested in contributing to this. Let me know if you'd like to discuss the implementation approach.

*(Connected through [Echeo](https://echeo.io))*

**Comment by politas180** [comment]
# Anti-Squatting System Proposal for gno.land

## Executive Summary

I propose a **three-tier namespace architecture** combined with **economic mechanisms** and **decentralized governance**, implemented as a **pluggable `p/antisquat` library**.

## Core Architecture

### Three-Tier Namespace System

| Tier | Format | Cost | Purpose |
|------|--------|------|---------|
| **Tier 1: Provisional** | `{name}-{hash6}` | Gas only | Permissionless entry, 3 per address |
| **Tier 2: Verified** | `{name}` | Length-based fee | Verified users (activity/stake/external ID) |
| **Tier 3: Premium** | `{short}` (1-3 chars) | Vickrey auction | Scarce, high-value names |

### Key Mechanisms

**1. Economic Disincentives**
- Length-based pricing (ENS-style): 3-char = 1000 GNOT/yr, 5+ char = 50 GNOT/yr
- Annual renewal requirement
- 10% transfer fee to discourage speculation

**2. Vickrey Sealed-Bid Auctions** (for Tier 3)
- Commit phase (7 days): Submit `hash(bid, salt, address)`
- Reveal phase (3 days): Reveal actual bids
- Winner pays second-highest price
- Prevents front-running and bid manipulation

**3. Dutch Auctions** (for expired names)
- Start at 100x last renewal price
- Decay to floor over 30 days
- First to claim at current price wins

**4. Governance Integration** (per #3020)
- Dispute resolution through Moderation DAO
- Reserved names registry (system, admin, gno, etc.)
- Trademark fast-track verification

**5. External Identity Integration** (per #4937)
- Format: `{username}@{provider}` (e.g., `kouteki@github`)
- OAuth verification through trusted oracles
- Inherently anti-squat (tied to external identity)

## p/antisquat Library Design

```gno
package antisquat

type Config struct {
    EnableProvisional  bool
    EnableVerified     bool
    EnablePremium      bool
    MaxProvisionalPerAddress int
    PricingTable       map[int]int64
    ModerationDAOAddress std.Address
    // ...
}

type AntiSquat struct { /* ... */ }

func New(config *Config) *AntiSquat
func (as *AntiSquat) RegisterProvisional(name string) (string, error)
func (as *AntiSquat) RegisterVerified(name string, method VerificationMethod) error
func (as *AntiSquat) StartAuction(name string) (*VickreyAuction, error)
func (as *AntiSquat) FileDispute(name string, dtype DisputeType, evidence string) (uint64, error)
func (as *AntiSquat) Resolve(name string) (std.Address, error)
```

## How This Addresses Requirements

| Requirement | Solution |
|-------------|----------|
| Prevent multiple registrations | 3 provisional limit + economic cost |
| Safeguard precious names | Reserved names registry (DAO-controlled) |
| Manage trademark conflicts | Dispute resolution + fast-track verification |
| Pluggable as p/ library | Full `p/antisquat` package design |
| No DNS reliance | Activity-based + stake-based + external identity verification |

## Verification Methods for Tier 2

1. **Proof-of-Activity**: 50+ txs, 90+ days on-chain, 1+ deployed realm
2. **Stake-Based**: Lock GNOT proportional to name length for 1 year
3. **External Identity**: OAuth verification through GitHub/Twitter/Discord

## Migration Path

1. Deploy `p/antisquat` alongside existing `r/sys/users`
2. 6-month migration period with priority claims for existing users
3. Old names remain resolvable through alias system
4. Full transition with upgrade path from Tier 1 → Tier 2

## Security Considerations

- **Front-running prevention**: Commit-reveal scheme
- **Sybil resistance**: Economic cost + activity requirements + external verification
- **Oracle trust**: Multi-sig quorum for external identity verification
- **Emergency controls**: Governance-controlled circuit breakers

## Full Proposal

I've prepared a detailed 500+ line proposal with complete implementation specifications, including:
- Full Gno code examples for all mechanisms
- Auction contract implementations
- Dispute resolution workflows
- Reserved names management
- Fee distribution model

**Full proposal**: Available on request or in a gist/PR

---

I'd love to discuss this approach and iterate based on feedback. This design synthesizes ideas from @thinhnx-var's sealed bidding proposal, @miguelito766's two-tier concept, and @AnhVAR's verified/unverified distinction, while adding economic mechanisms inspired by ENS.

Ready to implement once approach is validated.

**Comment by bimakw** [comment]
Hey, I'd like to work on this one.

Went through the codebase - current anti-squatting is basically just the 1 GNOT fee in `r/gnoland/users/v1/users.gno:37`, which doesn't scale. @politas180's proposal is comprehensive but I think we can ship something leaner first.

My plan:
- Start with `p/gnoland/antisquat` as a pluggable library (per @moul's suggestion)
- Core: rate limiting (3/address/30d) + protected names AVL tree + length pricing
- Later: sealed bid auction for premium names

I'll drop a design doc PR first before coding. One question - for the protected names list, should it live in the antisquat package itself or better as a separate `p/gnoland/protectednames` that antisquat imports?

**Comment by politas180** [comment]
Following up on my earlier proposal — I've implemented a working `p/gnoland/antisquat` package with tests.

  **PR:** https://github.com/gnolang/gno/pull/5126

  What's included:
  - Rate limiting (per-address, O(log n) via AVL trees)
  - Protected names registry (dynamic add/remove for GovDAO)
  - Length-based pricing (correct ugnot: 1 GNOT = 1,000,000 ugnot)
  - Name lifecycle (register, renew, resolve, expire + grace)
  - 20+ unit tests
  - Design document with Phase 1 + Phase 2 roadmap

  I reviewed the feedback on #5074 and proactively addressed all six concerns (pricing, performance, pruning, dynamic
  protected names, auction penalties, fee handling).

  Ready to iterate based on feedback.

**Comment by lmee** [comment]
I reviewed "Anti-squatting System for gno.land namespaces" and read all comments before proposing a patch path.

Problem I see: We need a generic anti-squatting system for r/sys/users to set up an automatic auction system. This will prevent squatting and ensure fair access. Initially, r/sys/users will remain invite-only.
Root-cause hypothesis: an FFI boundary mismatch (types/ownership/error handling) between wrapper code and native APIs.
Useful clue from comments: Hi, I'm not sure i got all the points but here's what i propose : EDIT : I think i totally missunderstood the subject. Can you please detail what kind of behavior you would like to.
First patch plan: define a minimal wrapper surface first, align types and ownership semantics with the native API, then add deterministic tests for correctness, memory safety, and failure-path behavior.
If this direction matches what you want, I can publish repro notes first so review starts from concrete evidence.

**Comment by huangpei2026psq** [comment]
GitHub username: huangpei2026psq
I'm interested in this $16 bounty.
Please let me know the next steps. Thanks!

**Comment by MrLawrenceKwan** [comment]
I’d like to submit a **design plan** (no implementation yet) for #2727.

## Goal
Build a generic, importable anti-squatting system for `r/sys/users` with automatic auction flow for high-risk names, while keeping v1 invite-only behavior intact.

## Design overview
I propose a reusable realm/library:
- `p/sys/namerules` (pure rule engine, scoring, normalization)
- `r/sys/nameguard` (stateful policy + auction orchestration)

`r/sys/users` calls into `r/sys/nameguard` before any registration is finalized.

---

## 1) Name classification pipeline
Input name goes through deterministic stages:

1. **Normalize**
   - lowercase
   - trim spaces
   - NFC normalization (if available in Gno stack; otherwise strict ASCII for v1)

2. **Validate base policy**
   - existing username constraints from `r/sys/users`
   - reserved protocol words blocked (DAO-managed set)

3. **Risk scoring** (0-100)
   - exact reserved/protected list match => reject
   - short/common names (e.g. length <= 4) => high score
   - high-confusability patterns (lookalike chars / obvious homograph classes) => high score
   - brand/known-public-entity dictionary hit => high score
   - entropy / random suffix heuristics => lower score

4. **Decision bands**
   - LOW risk: immediate register allowed
   - MEDIUM risk: delayed claim window + challenge period
   - HIGH risk: must go through auction

The thresholds are DAO-configurable.

---

## 2) Auction model (for HIGH risk names)
State machine per name:

`Idle -> PendingAuction -> ActiveAuction -> Settled | Cancelled`

### Auction mechanics
- sealed-bid commit/reveal (to reduce sniping)
- bid bond required
- reveal timeout with slashing of non-revealed commits
- winner gets right-to-register token/record for limited period
- if winner does not finalize, fallback to next valid bidder or restart

### Anti-abuse
- per-address concurrent auction limits
- minimum bond by risk tier
- cooldown on failed finalization
- challenge bond for disputes

---

## 3) Generic interfaces (pseudo API)
```gno
// p/sys/namerules
func Normalize(name string) string
func Score(name string, ctx RuleContext) ScoreResult

// r/sys/nameguard
func CheckEligibility(caller address, name string) (Decision, error)
func StartAuction(name string) (auctionID uint64, error)
func CommitBid(auctionID uint64, commitment string, bond coins) error
func RevealBid(auctionID uint64, amount coins, nonce string) error
func FinalizeAuction(auctionID uint64) (winner address, error)
func ConsumeWinRight(name string, winner address) error
```

`r/sys/users.RegisterUser(...)` flow:
1) call `CheckEligibility`
2) if LOW => proceed
3) if HIGH => require `ConsumeWinRight` proof before write

---

## 4) Storage model
- `NamePolicy` (thresholds, reserved sets version)
- `NameRiskCache` (optional deterministic cache)
- `AuctionsByName`
- `AuctionBidsCommit` / `AuctionBidsReveal`
- `WinRights` (name -> winner, expiry)
- `Disputes`

All critical transitions emit events for indexers/gnoweb.

---

## 5) Governance / upgradeability
- policy params, protected lists, and thresholds managed by GovDAO proposals
- emergency pause for auction creation (not settlement)
- migration-friendly versioned structs

---

## 6) Rollout plan
1. **Phase A (MVP anti-squatting gate):** scoring + allow/reject + delayed claims (no auction yet)
2. **Phase B:** auction commit/reveal for HIGH risk names
3. **Phase C:** dispute hooks + richer confusability dictionaries + analytics feedback loop

---

## 7) Why this fits #2727
- generic/importable library+realm (not hardcoded into users)
- automatic auction path for suspicious names
- compatible with current invite-only `r/sys/users` transition strategy
- DAO-governable and forward-compatible with #2827 v2 vision

If this direction is acceptable, I can proceed with a minimal, scoped implementation PR in phases (A first), modifying only the relevant `r/sys/users` + new `nameguard/namerules` modules.

**Comment by thehowl** [comment]
Hi. We identified that over the past weeks, the bounty issues have attracted a steady stream of low-quality submissions using automated artificial intelligence tools. We are unable to currently process through these issues and don't consider these to be a valuable submission to our development community. As a consequence, issues and PR's related to the bounty program are no longer eligible. We'll update the contribution page on the gno.land website to reflect this. Contributors who contributed PR's that addressed bounties before September 2025 may still be eligible if there was a comment by a core team member specifically approving/validating the contribution.
